### PR TITLE
[READY TO MERGE] Improve docs for Multinomial and Categorical distributions

### LINKED ISSUE #7014: multinomial with inf fails in cuda
`torch.multinomial` throws a `RuntimeError` if the probability distribution includes `float('inf')`. On CPU there is no similar problem. This can happen if you take a (what is probably naive) approach to sampling logits like this:

```python
def sample_temp(logits, temperature):
    probs = torch.exp(logits / temperature)
    return torch.multinomial(probs, 1)
```

It's more likely to cause a problem if your logits are very large or your temperature is very small.

## Code example

```python
import torch
x = torch.Tensor([0, float('inf')])
print(torch.multinomial(x, 1))
print(torch.multinomial(x.to('cuda'), 1))
```

```python-traceback
tensor([ 1])
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-98dbbe29edf5> in <module>()
      2 x = torch.Tensor([0, float('inf')])
      3 print(torch.multinomial(x, 1))
----> 4 print(torch.multinomial(x.to('cuda'), 1))

RuntimeError: cuda runtime error (59) : device-side assert triggered at /opt/conda/conda-bld/pytorch_1524590031827/work/aten/src/THC/generic/THCTensorCopy.c:20
```

## System Info

```
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.1.85

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 390.48
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a
/usr/local/cuda-9.1/lib64/libcudnn.so.7.1.2
/usr/local/cuda-9.1/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.14.2)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] cuda80                    1.0                           0    soumith
[conda] cuda91                    1.0                  h4c16780_0    pytorch
[conda] pytorch                   0.4.0           py36_cuda9.1.85_cudnn7.1.2_1  [cuda91]  pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
```

## Back story

It may look simple, but this was a particularly difficult bug to track down :) I moved my RNN sampling from CPU (with numpy) to GPU (with torch) and started getting errors like:

* `RuntimeError: CUDNN_STATUS_EXECUTION_FAILED`
* `RuntimeError: cuda runtime error (59) : device-side assert triggered at /opt/conda/conda-bld/pytorch_1524590031827/work/aten/src/THC/generic/THCTensorCopy.c:20`
* `RuntimeError: cuda runtime error (59) : device-side assert triggered at /opt/conda/conda-bld/pytorch_1524590031827/work/aten/src/THC/generated/../THCReduceAll.cuh:339`

I saw https://github.com/pytorch/pytorch/issues/5213 and https://github.com/pytorch/pytorch/issues/6306 had a similar error, which made me think it was related to my sampling code not moving data around correctly, but this didn't explain the intermittent behavior. I audited that and continued searching. Then I saw https://github.com/pytorch/pytorch/issues/5561 which made me check that my embedding indices were being set correctly (not out of bounds). So I looked closer into multinomial and found this bug.

One way to get around the sampling bug above is to just do everything on the CPU. Another solution is to use Gumbel-softmax style sampling:

```python
def sample(logits, temperature=1.0):
    if temperature == 0:
        return torch.argmax(logits, dim=-1)
    noise = torch.FloatTensor(logits.shape).to(logits.device)
    noise.uniform_(1e-5, 1-1e-5)
    return torch.argmax(logits / temperature - torch.log(-torch.log(noise)), dim=-1)
```

But unfortunately this takes longer to execute than the first round of sampling code.

### PR DESCRIPTION
**Pull request by yf225**
Closes https://github.com/pytorch/pytorch/issues/7014 by specifying the constraints on the input for Multinomial and Categorical distributions.

### CODE REVIEW

File: torch/distributions/multinomial.py [hunk 1]
```diff
@@ -15,7 +15,8 @@ class Multinomial(Distribution):
     Note that `total_count` need not be specified if only :meth:`log_prob` is
     called (see example below)
 
-    .. note:: :attr:`probs` will be normalized to be summing to 1.
+    .. note:: :attr:`probs` must be non-negative, not infinite and have a non-zero sum,
+    and it will be normalized to be summing to 1.
```
Comment (li-roy) [inline torch/distributions/multinomial.py]: Maybe "normalized to sum to 1"?

**Review by li-roy — APPROVED** [review]

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -2725,7 +2725,8 @@ def parse_kwargs(desc):
 
 .. note::
     The rows of :attr:`input` do not need to sum to one (in which case we use
-    the values as weights), but must be non-negative and have a non-zero sum.
+    the values as weights), but must be non-negative, not infinite and have
```
Comment (SsnL) [inline torch/_torch_docs.py]: nit: people usually say "finite" instead of "not infinite"

**Review by ssnl — APPROVED** [review]
lgtm

**Review by soumith — APPROVED** [review]

**Comment by guotong1988** [comment]
Thank you.

**Pull request merged**